File: MatlabCode/02_b_Homework_MonteCarlo/wind_turbine.m
%% Clean workspace
close all;
clear all;
clc;



%% Parameters

% Number of days to simulate
N=365;

% Number of samples per day
Ns=24; % sampling time = 1 hour



%% Simulate wind speed measurements

% Use Weibull(lambda,kappa)
lambda=8;
kappa=1.6;
v=wblrnd(lambda,kappa,N*Ns,1);



%% Simulate energy generation

% Allocate vector to store energy
e=zeros(N*Ns,1);

% Simulate generated energy [kWh]
for k=1:N*Ns
    e(k)=characteristic_curve(v(k));
end

% Convert energy to [MWh] (for readability)
e=e/1000;



%% Compute and display results

% Arrange energy in days
% Each row of matrix "energy" is a day
% Each column of matrix "energy" is an hour
energy=reshape(e,Ns,N)';

% Compute daily generated energy
% Each component of vector G is a day of generated energy
G=sum(energy,2);

% mean
Gm=mean(G);

% std dec
Gs=std(G);

% plot input data
figure
ax1=subplot(2,1,1);
plot(e,'r')
grid on
xlabel('time [hours]')
ylabel('energy [MWh]')
title('Generated energy')
ax2=subplot(2,1,2);
plot(v)
grid on
xlabel('time [hours]')
ylabel('wind speed [m/s]')
title('Wind speed')
% link x-axis of two subplots, useful when zooming
linkaxes([ax1,ax2],'x');

% pdf 
figure
histogram(G,'Normalization','pdf')
grid on
xlabel('Daily energy [MWh]')
ylabel('pdf')
title('Histogram')


% cdf
[F, x]=ecdf(G);
figure
plot(x,F)
grid on
xlabel('Daily energy [MWh]')
ylabel('cdf')
title('Empirical cdf')

% exceedance probabilities
%
% P90:   P(G>=P90) = 0.9
%  ->  1-P(G>=P90) = 0.1
%  ->    P(G< P90) = 0.1
%
% Hence, P90 is the 10-th percentile
%
P90 = prctile(G,10);
P75 = prctile(G,25);
P50 = prctile(G,50);


% write some results
disp(['    Number of days: ' num2str(N)])
disp(['Number samples/day:  ' num2str(Ns)])
disp(' ')
disp('    Daily energy G   [MWh]')
disp('  -----------------------------')
disp(['              mean: ' num2str(Gm)])
disp(['           std dev: ' num2str(Gs)])
disp(' ')
disp(['               P90: ' num2str(P90)])
disp(['               P75: ' num2str(P75)])
disp(['               P50: ' num2str(P50)])

File: PythonCode/02_b_Homework_MonteCarlo/wind_turbine.py
import numpy as np
from characteristic_curve import characteristic_curve
import matplotlib.pyplot as plt

N = 365                 # number of days to simulate
Ns = 24                 # samples per day (1/hour)

# simulate wind speed measurements
l = 8                   # lambda parameter
k = 1.6                 # kappa parameter

v = np.random.weibull(k, size=(N*Ns))*l

# simulate energy generation
e = np.array([])
for vh in v:
    e = np.append(e, characteristic_curve(vh))

# convert energy to [MWh]
e /= 1000

# display results
# each row of matrix energy is a day, each column is an hour
energy = np.reshape(e, (N, Ns))

# compute daily generated energy (sum along rows)
G = np.sum(energy, axis=1)

# mean and std
Gm = np.mean(G)
Gs = np.std(G)

# plot input data
fig = plt.figure(figsize=(5, 10))
subfig = fig.add_subplot(2, 1, 1)
subfig.set_title('Generated energy')
subfig.set(xlabel='Time [hours]', ylabel='Energy[MWh]')
plt.plot(e, 'r')
plt.grid()
subfig = fig.add_subplot(2, 1, 2)
subfig.set_title('Wind speed')
subfig.set(xlabel='Time [hours]', ylabel='Wind speed [m/s]')
plt.plot(v)
plt.grid()
plt.show()

# pdf
fig, ax = plt.subplots()
fig.suptitle('Histogram')
ax.set(xlabel='Daily energy [MWh]', ylabel='pdf')
plt.hist(G, bins='auto', density=True)
plt.grid()
plt.show()

# cdf
fig, ax = plt.subplots()
fig.suptitle('Empirical cdf')
ax.set(xlabel='Daily energy [MWh]', ylabel='cdf')
plt.hist(G, bins='auto', density=True, cumulative=True)
plt.grid()
plt.show()

# exceedance probabilities
#
# P90:   P(G>=P90) = 0.9
#  ->  1-P(G>=P90) = 0.1
#  ->    P(G< P90) = 0.1
#
# percentiles stores respectively the 10-25-50 percentile
# P90 is the 10th percentile

percentiles = np.percentile(G, [10., 25., 50.])

# write some results
print(f'Number of days: {N}')
print(f'Samples/day: {Ns}')
print('Daily Energy [MWh]')
print('-'*20)
print(f'Mean: {Gm}')
print(f'std dev: {Gs}')
print()
print(f'P90: {percentiles[0]}')
print(f'P75: {percentiles[1]}')
print(f'P50: {percentiles[2]}')
